plugins/inputs/http_listener_v2/http_listener_v2.go

```diff
@@ -0,0 +1,304 @@
+  basic_username = "foobar"
+  basic_password = "barfoo"
+
+  ## Data format to consume.
+  ## Each data format has its own unique set of configuration options, read
+  ## more about them here:
+  ## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+  data_format = "influx"
+`
+
+func (h *HTTPListenerV2) SampleConfig() string {
+	return sampleConfig
+}
+
+func (h *HTTPListenerV2) Description() string {
+	return "Generic HTTP write listener"
+}
+
+func (h *HTTPListenerV2) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (h *HTTPListenerV2) SetParser(parser parsers.Parser) {
+	h.Parser = parser
+}
+
+// Start starts the http listener service.
+func (h *HTTPListenerV2) Start(acc telegraf.Accumulator) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if h.MaxBodySize == 0 {
+		h.MaxBodySize = defaultMaxBodySize
+	}
+
+	if h.ReadTimeout.Duration < time.Second {
+		h.ReadTimeout.Duration = time.Second * 10
+	}
+	if h.WriteTimeout.Duration < time.Second {
+		h.WriteTimeout.Duration = time.Second * 10
+	}
+
+	h.acc = acc
+
+	tlsConf, err := h.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	server := &http.Server{
+		Addr:         h.ServiceAddress,
+		Handler:      h,
+		ReadTimeout:  h.ReadTimeout.Duration,
+		WriteTimeout: h.WriteTimeout.Duration,
+		TLSConfig:    tlsConf,
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", h.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", h.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	h.listener = listener
+	h.Port = listener.Addr().(*net.TCPAddr).Port
+
+	h.wg.Add(1)
+	go func() {
+		defer h.wg.Done()
+		server.Serve(h.listener)
+	}()
+
+	log.Printf("I! Started HTTP listener V2 service on %s\n", h.ServiceAddress)
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (h *HTTPListenerV2) Stop() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	h.listener.Close()
+	h.wg.Wait()
+
+	log.Println("I! Stopped HTTP listener V2 service on ", h.ServiceAddress)
+}
+
+func (h *HTTPListenerV2) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if req.URL.Path == h.Path {
+		h.AuthenticateIfSet(h.serveWrite, res, req)
+	} else {
+		h.AuthenticateIfSet(http.NotFound, res, req)
+	}
+}
+
+func (h *HTTPListenerV2) serveWrite(res http.ResponseWriter, req *http.Request) {
+	// Check that the content length is not too large for us to handle.
+	if req.ContentLength > h.MaxBodySize {
+		tooLarge(res)
+		return
+	}
+
+	// Check if the requested HTTP method was specified in config.
+	isAcceptedMethod := false
+	for _, method := range h.Methods {
+		if req.Method == method {
+			isAcceptedMethod = true
+			break
+		}
+	}
+	if !isAcceptedMethod {
+		methodNotAllowed(res)
+		return
+	}
+
+	// Handle gzip request bodies
+	body := req.Body
+	if req.Header.Get("Content-Encoding") == "gzip" {
+		var err error
+		body, err = gzip.NewReader(req.Body)
+		defer body.Close()
```

Comment:
If an error occurred, body will probably be nil and calling Close might panic, move this under the err check.